### DISCUSSION TITLE
Slow Inference speed on an updated model

### DISCUSSION THREAD

**Discussion by mehulrastogi** [talmolab/sleap#1442] [Help!]
I am facing problems with slow inference speeds. I recently updated my model to add more training labels, but the inference speed on the retrained model seemed to be very slow (around 30fps) when compared to the original model (around 90fps).  This is even though that the processes are utilising the GPU during inference.

I have tried retraining both the models (with same configuration with different drivers and GPUs) and am able to reproduce the results. The only difference being the amount of labels. Original one had 370 labelled frames vs the new one has around 420 labelled frames. Although I am highly sceptical that this is the reason for the slow down and need help on debugging this issue. 

The details for the setup I have used for training:-

GeForce RTX 3070
![image](https://github.com/talmolab/sleap/assets/22076299/cee01b33-489a-4fdd-bc2f-a532c8773b00)

I have also tried training on a cluster using the following GPUs with similar results. 

first:
![image](https://github.com/talmolab/sleap/assets/22076299/eb31992f-68b4-4c8e-a319-16def9f8eeed)

second: 
![image](https://github.com/talmolab/sleap/assets/22076299/f3e83509-5c99-4b97-929f-862b6200c190)

I am currently using the following configuration files to train the multi animal topdown-id model. 

[octagon_multi.centered_instance_multiclass_cfg_v15_upload.txt](https://github.com/talmolab/sleap/files/12301306/octagon_multi.centered_instance_multiclass_cfg_v15_upload.txt)

[octagon_multi.centroid_cfg_v15_upload.txt](https://github.com/talmolab/sleap/files/12301307/octagon_multi.centroid_cfg_v15_upload.txt)

Some general questions:- 
- I have a use case to do inferences on historical data (some 700 video files). It is possible that the inference happens on a different GPU with different drivers than the one used to train the model. Will this cause some problems with the speed or the inference accuracies?
- What is the best recommended CUDA and nvidia drivers to use to train SLEAP models? 

Thanks A Lot

**Comment by roomrys** [comment]
Hi @mehulrastogi,

We have not noticed that driver versions will _significantly_ affect speed... we usually recommend the most up-to-date drivers, but the constraint is really just the oldest version which runs. I am wondering if you are on different versions of SLEAP between training the fast/slow models?

Thanks,
Liezl

**Reply by mehulrastogi** [reply]
Hi @roomrys. Thank you for your reply 

I also wondered the same initially. But I am able to replicate the issue with the latest sleap version (reinstalled sleap in the conda environment). 

I can try training a new network with the recommended set of parameters on a particular GPU on the latest drivers. What according to you would be a good set of parameters for the A4500s? The parameters I am training with are there with the original post. I can also enable any logs you might require to debug the issue

**Comment by roomrys** [comment]
Hi @mehulrastogi,

I know you said you retrained the model... Are all the parameters in the `training_config.json`s between models the same?

One thing that may slow down the model is the input scaling on the centroid model. In the parameters you sent, it had an input scaling of 1.0 which may require more parsing over the image to find the centroid than with a smaller input scaling.

Thanks,
Liezl

**Reply by mehulrastogi** [reply]
I see, thanks for pointing this out. Scaling it down and training it. Also, it seems like TensorFlow prefers a particular combination of CUDA and cuDNN:-
- https://docs.nvidia.com/deeplearning/cudnn/support-matrix/index.html
- https://www.tensorflow.org/install/source#gpu

Did a clean install. Hopefully this works.
